_posts/2023/2023-02-27-consider-find_each-for-active-record-assocation-loops.md

```diff
@@ -0,0 +1,62 @@
+---
+title: "Consider find_each for looping through Active Record associations"
+description: "Rarely used, but protects you against foot shooting"
+layout: article
+category: ruby
+image:
+  base: '2023/consider-find_each-for-looping-through-active-record-associations'
+  alt: "Light loops"
+  credit: "Nareeta Martin"
+  source: "https://unsplash.com/photos/za2gMSfLmzU"
+
+---
+
+The standard method for enumerating through groups of objects is `each`, both through arrays in Ruby and through Active Record models in Rails.
+
+However, if you are looping over all the records for a model (perhaps to backfill data) or even over a large association you can encounter severe memory and speed issues both loading and processing enormous amounts of records as you loop.
+
+You should consider the functionality provided by [Active Record’s Batches](https://api.rubyonrails.org/classes/ActiveRecord/Batches.html), we’ll demonstrate using `find_each`.
+
+
+## Instead of…
+
+…looping through lots of Active Record objects using `.each`:
+
+```ruby
+post.comments.each do |comment|
+  # Do stuff with each comment: enqueue a job
+end
+```
+
+
+## Use…
+
+…use `.find_each` to more efficiently load records from the database:
+
+```ruby
+post.comments.find_each do |comment|
+  # Do stuff with each comment: enqueue a job?
+end
+```
+
+
+## Why?
+
+Using `each` makes _one_ SQL call to the database and tries to load the entire set of objects into memory and then loop over them. It’s the same as calling `post.comments.all.each` in the first example.
+
+This is problem in two dimensions. Firstly the database query may take a long time to execute or may timeout. Secondly, when (and if) it does, there’s likely to be significant memory usage as it loads all the records into memory in order to loop over them. 
+
+Using `.find_each` makes a series of more efficient SQL queries (with a bunch of sensible defaults) and uses a cursor-based approach to retrieve records from the database, which can be more efficient than loading all records into memory at once.
+
+
+## Why not?
+
+At a certain size you’ll need to move on from looping over every record.
```

Comment:
This isn't a "Why not?", right?

This is perhaps for the section above? Almost like: well, at a certain point you'll _have_ to do this!

Reply:
No, this is because even _this_ won't work once you're at a certain scale!

Reply:
```suggestion
At large scale you’ll need to use even more advanced techniques and you’ll have to move beyond long running loops over every record. Perhaps to reduce a high volume of read/write operations on your database, perhaps to try and get this loop to run in minutes rather than hours!
```